File: bst100.py
import time

## Recursive build function, it takes a number and places it in appropriate array location
## Parent at n, left child at 2n +1, and right child at 2n+2
def build_tree(bst_array, number, current_indeks):
    if(bst_array[current_indeks]==' '):
        bst_array[current_indeks] = number
        current_indeks = 0
    elif(number < bst_array[current_indeks]):
        build_tree(bst_array, number, 2*current_indeks + 1)
    elif(number > bst_array[current_indeks]):
        build_tree(bst_array, number, 2*current_indeks + 2)

## Recursive print function, prints tree as inorder traversal and writes to a file in append mode
def print_tree_inorder_recursive(sorted_file, bst_array, local_indeks):
    if(bst_array[local_indeks] == " "):
       return False
    print_tree_inorder_recursive(sorted_file, bst_array, 2*local_indeks + 1)
    print(bst_array[local_indeks])
    sorted_file.write(str(bst_array[local_indeks]) + " ")
    print_tree_inorder_recursive(sorted_file, bst_array, 2* local_indeks + 2)

number_list = []

## Read numbers through number_list array
number_file = open('numbers100.txt', mode = "r")
for line in number_file:
    splitted = line.split()
    for a_one in splitted:
        a_one = a_one.strip()
        number_list.append(int(a_one))

number_file.close()

bst_array = []

t_before = time.time()  ## Measure the start time

## Initialize the tree array, I don't know if there is a more efficient way
for i in range(50000):
    bst_array.append(" ")

## For every number in number_list place the number in appropriate index to bst_array
for i in range(len(number_list)):
    build_tree(bst_array, number_list[i], 0)

t_after = time.time()  ## Measure the end time

## Open the file in appending mode
sorted_number_file = open("sorted_100_py.txt" , mode = "a+")

## Print the tree and write to file in function
print_tree_inorder_recursive(sorted_number_file, bst_array, 0)

## Execution time
print(t_after-t_before)

File: bst_100.c
// EMRE KILAVUZ
// 220201019

//Because of the program reads 500 numbers, the name of the file is bst_500.c

#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <time.h>

char* trim(char* token);
void build_tree(int bst_array[50000], int number, int current_indeks);
void print_tree_inorder_recursive(FILE* sorted_file_pointer, int bst_array[50000], int local_indeks);

int main(){
    clock_t time_start , time_end;  // Define time variables
    double bst_build_time; 
    int numbers[100];   // Define input number array
    FILE* file_pointer;
    char *str = malloc(125*sizeof(char));  // For file reading and tokenization
    char* file_name = "./numbers4.txt";
    char *token;
    int counter = 0;
    file_pointer = fopen(file_name, "r");
    if (file_pointer == NULL){
        printf("Program could not open this file %s",file_name);
        return 1;
    }
    // Read 125 character each time, tokenize, convert integer and place it into array
    while(fgets(str,125, file_pointer)!=NULL){
        token = strtok(str," ");
        while(token != NULL){
            char* trimmed = trim(token);
            if(trimmed[0] != '\0'){
                int number = atoi(trimmed);
                numbers[counter] = number;
                counter += 1;
            }
            
            token =strtok(NULL," ");
        }
    }
    fclose(file_pointer); // Finish reading file, close file


    time_start = clock();  // Measure start time

    // Initialize tree array, fill minus one, as an exception number
    int bst_array[50000];
    for(int i=0; i< 50000; i++){
        bst_array[i] = -1;
    }
    // Place the input numbers into tree array
    for(int j=0; j<100;j++){
        build_tree(bst_array, numbers[j], 0);
    }
    
    time_end = clock();  // Measure the end time
    bst_build_time = ((double) (time_end - time_start)) / CLOCKS_PER_SEC; // BST build time
    FILE* file_pointer_write;
    // Open file in append mode
    file_pointer_write = fopen("sorted_numbers_100_c.txt", "a");
    
    // Print the tree as inorder traversal and write to file in recursive function
    print_tree_inorder_recursive(file_pointer_write, bst_array, 0);
    fclose(file_pointer_write);
    printf("%lf\n", bst_build_time);
    return 0;
}

// Print the tree using inorder traversal and writes to file in a recursive manner
void print_tree_inorder_recursive(FILE* sorted_file_pointer, int bst_array[50000], int local_indeks){
    if(bst_array[local_indeks] == -1){
        return;
    }
    print_tree_inorder_recursive(sorted_file_pointer, bst_array, 2*local_indeks +1);
    printf("%d ", bst_array[local_indeks]);
    char* snum = malloc(7*sizeof(char));
    for(int l=0; l<7; l++){
        snum[l] = '\0';
    }
    sprintf(snum, "%d ", bst_array[local_indeks]);
    //strcat(snum, " ");
    fputs(snum, sorted_file_pointer);
    print_tree_inorder_recursive(sorted_file_pointer, bst_array, 2*local_indeks +2);
}

// Parent at n, left child at 2n +1 and right child at 2n+2
void build_tree(int bst_array[50000], int number, int current_indeks){
    if(bst_array[current_indeks] == -1){  // minus 1 is the exception number in tree
        bst_array[current_indeks] = number;
        current_indeks = 0;
    }
    else if(number < bst_array[current_indeks]){
        build_tree(bst_array, number, 2*current_indeks +1);
    }
    else if(number > bst_array[current_indeks]){
        build_tree(bst_array, number, 2*current_indeks +2);
    }
}

// Remove white spaces, new lines and tabs
char* trim(char* token){
    char* trimmed = malloc(50 * sizeof(char));
    int trim_count1 = 0;
    int trim_count2 = 0;

    while(token[trim_count1] != '\0'){
        if(token[trim_count1] != ' ' && token[trim_count1] != '\n' && 
                token[trim_count1]!='\t' && token[trim_count1]!='\r')
                {
            trimmed[trim_count2] = token[trim_count1];
            trim_count2++;
        }
        trim_count1++;
    }
    trimmed[trim_count2] = '\0';
    return trimmed;
}

